### PR TITLE
feat: add underlying error info to reported NSErrors

### PR DESCRIPTION
**Pull request by armcknight**
## :scroll: Description

Refactor SentryNSError to accept just the NSError instance in init, and don't expose properties that aren't used outside of its implementation.

Then, wrap all underlying errors in SentryNSError and serialize them.

## :bulb: Motivation and Context

https://github.com/getsentry/sentry-cocoa/issues/3063

## :pencil: Checklist

You have to check all boxes before merging:

- [ ] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

**Review by brustolin — COMMENTED** [review]
According to the [docs](https://develop.sentry.dev/sdk/event-payloads/exception/#examples) the correct way to report underlying errors is to add them to the event exception array.
The [Exception Interface](https://develop.sentry.dev/sdk/event-payloads/exception/) has no `underlying_errors` property.

**Comment by armcknight** [comment]
> According to the [docs](https://develop.sentry.dev/sdk/event-payloads/exception/#examples) the correct way to report underlying errors is to add them to the event exception array. The [Exception Interface](https://develop.sentry.dev/sdk/event-payloads/exception/) has no `underlying_errors` property.

Thanks for pointing me to this @brustolin! The pertinent part seems to be:

> the exception attribute may be a flat list of objects...Multiple values represent chained exceptions and should be sorted oldest to newest

**Comment by armcknight** [comment]
One problem though @brustolin is that this is for SentryNSError, and you're talking about something pertaining to SentryException; these are different API. I'll look at more docs for errors and see if we have something regarding underlying error reporting.

**Comment by armcknight** [comment]
So I found this [here](https://develop.sentry.dev/sdk/event-payloads/), which I think is why I figured I could get away with adding a key that isn't already specified: 
image

File: Sources/Sentry/Public/SentryEvent.h [hunk 1]
```diff
@@ -138,8 +138,8 @@ NS_SWIFT_NAME(Event)
 @property (nonatomic, strong) NSArray<SentryThread *> *_Nullable threads;
 
 /**
- * General information about the @c SentryException, usually there is only one
- * exception in the array.
+ * General information about the @c SentryException. Multiple exception indicate a chain of
+ * exception encountered, starting with the oldest at the beginning of the array.
```
Comment (armcknight) [inline Sources/Sentry/Public/SentryEvent.h:142]:
```suggestion
 * General information about the @c SentryException. Multiple exceptions indicate a chain of
 * exceptions encountered, starting with the oldest at the beginning of the array.
```

File: Tests/SentryTests/Protocol/SentryNSErrorTests.swift [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import SentryTestUtils
```
Comment (armcknight) [inline Tests/SentryTests/Protocol/SentryNSErrorTests.swift:1]:
```suggestion
```

File: Tests/SentryTests/SentryClientTests.swift [hunk 3]
```diff
@@ -1522,29 +1531,28 @@ class SentryClientTest: XCTestCase {
...
-        
-        guard let mechanism = exception.mechanism else {
-            XCTFail("Exception doesn't contain a mechanism"); return
-        }
+
+        let mechanism = try XCTUnwrap(exception.mechanism)
+        let meta = try XCTUnwrap(mechanism.meta)
+        let actualError = try XCTUnwrap(meta.error)
         XCTAssertEqual("NSError", mechanism.type)
-        XCTAssertNotNil(mechanism.meta?.error)
```
Comment (armcknight) [inline Tests/SentryTests/SentryClientTests.swift:1545]: This gives us a more granular idea of which thing is missing; It is responsible for all of the `throws` additions in this file.

**Review by brustolin — APPROVED** [review]
LGTM

**Pull request merged**